File: lib/src/parser/xml_helper.dart
import 'package:musicxml_parser/src/exceptions/musicxml_structure_exception.dart';
import 'package:xml/xml.dart';

/// Utility class providing helper methods for parsing MusicXML documents.
///
/// This class includes functions for common XML operations such as retrieving
/// element line numbers, finding child elements, and converting element text
/// to various data types, tailored for the needs of MusicXML parsing.
class XmlHelper {
  /// Gets the line number of an XML [element], if available from its attributes.
  ///
  /// MusicXML elements do not standardly include line numbers. This method
  /// checks for a non-standard 'line' attribute that might be added by
  /// pre-processing tools or for debugging.
  ///
  /// Returns the line number if found, otherwise -1.
  static int getLineNumber(XmlElement? element) {
    if (element == null) return -1;

    final lineAttribute = element.getAttribute('line');
    if (lineAttribute != null) {
      final lineNum = int.tryParse(lineAttribute);
      if (lineNum != null) return lineNum;
    }
    return -1;
  }

  /// Finds the text content of an optional child element specified by a simple [path].
  ///
  /// The [path] is a '/' separated string representing a sequence of child elements.
  /// It also supports selecting an attribute using '@attributeName' as the last segment,
  /// or filtering by an attribute value using 'elementName[@attribute="value"]'.
  ///
  /// Example paths:
  /// - 'work/work-title' (finds `<work-title>` inside `<work>`)
  /// - 'identification/creator[@type="composer"]' (finds `<creator type="composer">` inside `<identification>`)
  /// - 'attributes/@version' (gets the 'version' attribute of the `<attributes>` element)
  ///
  /// Returns the trimmed text content of the found element or attribute, or `null` if not found.
  static String? findOptionalTextElement(XmlElement element, String path) {
    XmlElement? currentContextNode = element;
    final pathSegments = path.split('/');

    for (int i = 0; i < pathSegments.length; i++) {
      final segment = pathSegments[i];
      if (currentContextNode == null) return null;

      if (segment.startsWith('@')) {
        // Attribute selection should be the last segment
        if (i == pathSegments.length - 1) {
          final attributeName = segment.substring(1);
          return currentContextNode.getAttribute(attributeName);
        } else {
          // Attribute selection in the middle of a path is not supported by this simplified helper
          return null;
        }
      } else {
        // Element selection with optional predicate
        currentContextNode =
            _findElementFromSegment(currentContextNode, segment);
      }
    }

    return currentContextNode?.innerText.trim();
  }

  /// Internal helper to find a child element based on a [segment] name, handling predicates.
  ///
  /// A [segment] can be a simple element name (e.g., "note") or an element name
  /// with an attribute predicate (e.g., 'creator[@type="composer"]').
  ///
  /// Returns the first matching [XmlElement], or `null` if not found.
  static XmlElement? _findElementFromSegment(
      XmlElement parent, String segment) {
    final predicateMatch = RegExp(r'(.+?)\[(.+?)\]').firstMatch(segment);

    if (predicateMatch != null) {
      final elementName = predicateMatch.group(1);
      final predicate = predicateMatch.group(2);

      if (elementName == null || predicate == null) {
        // Malformed segment (e.g., "name[]" or "[]value"), treat as simple element name
        return parent.findElements(segment).firstOrNull;
      }

      // Handle @attribute="value" predicate
      final attributePredicateMatch =
          RegExp(r'@(.+?)="(.+?)"').firstMatch(predicate);
      if (attributePredicateMatch != null) {
        final attributeName = attributePredicateMatch.group(1);
        final attributeValue = attributePredicateMatch.group(2);

        if (attributeName != null && attributeValue != null) {
          try {
            return parent.findElements(elementName).firstWhere(
                  (el) => el.getAttribute(attributeName) == attributeValue,
                );
          } catch (e) {
            // firstWhere throws StateError if no element is found
            return null;
          }
        }
      }
      // Fallback for unhandled or malformed predicate: find first element by name if any
      return parent.findElements(elementName).firstOrNull;
    } else {
      // Simple element name, no predicate
      return parent.findElements(segment).firstOrNull;
    }
  }

  /// Finds and returns a required child [XmlElement] with the given [name].
  ///
  /// Throws a [MusicXmlStructureException] if the element is not found.
  /// The [requiredElement] parameter can be used to specify a more descriptive
  /// name in the exception message if [name] is too generic.
  static XmlElement getRequiredElement(
    XmlElement parent,
    String name, {
    String? requiredElement,
  }) {
    final elements = parent.findElements(name);
    if (elements.isEmpty) {
      throw MusicXmlStructureException(
        'Required element <${requiredElement ?? name}> not found as a child of <${parent.name.local}>.',
        requiredElement: requiredElement ?? name,
        parentElement: parent.name.local,
        line: getLineNumber(parent),
      );
    }
    return elements.first;
  }

  /// Finds and returns an optional child [XmlElement] with the given [name].
  ///
  /// Returns `null` if the element is not found.
  static XmlElement? findOptionalElement(XmlElement parent, String name) {
    final elements = parent.findElements(name);
    return elements.isNotEmpty ? elements.first : null;
  }

  /// Gets the trimmed text content of an [element] and parses it as an integer.
  ///
  /// Returns the integer value, or `null` if the [element] is `null`,
  /// its text is empty, or not a valid integer.
  static int? getElementTextAsInt(XmlElement? element) {
    if (element == null) return null;
    final text = element.innerText.trim();
    return int.tryParse(text);
  }

  /// Gets the trimmed text content of an [element] and parses it as a double.
  ///
  /// Returns the double value, or `null` if the [element] is `null`,
  /// its text is empty, or not a valid double.
  static double? getElementTextAsDouble(XmlElement? element) {
    if (element == null) return null;
    final text = element.innerText.trim();
    return double.tryParse(text);
  }

  /// Gets the trimmed text content of an [element] and interprets it as a boolean.
  ///
  /// Recognizes "yes" (case-insensitive) as `true` and "no" (case-insensitive)
  /// as `false`. For any other text, or if the element is `null` or its text is empty,
  /// returns the [defaultValue] (which is `false` if not specified).
  static bool getElementTextAsBool(XmlElement? element,
      {bool defaultValue = false}) {
    if (element == null) return defaultValue;
    final text = element.innerText.trim().toLowerCase();
    if (text == 'yes') return true;
    if (text == 'no') return false;
    return defaultValue;
  }

  /// Gets the value of an attribute by [attributeName] from an [element].
  ///
  /// Returns the attribute value as a `String?`, or `null` if the element is `null`
  /// or the attribute does not exist.
  static String? getAttributeValue(XmlElement? element, String attributeName) {
    if (element == null) return null;
    return element.getAttribute(attributeName);
  }

  /// Gets the value of an attribute by [attributeName] from an [element] and parses it as a double.
  ///
  /// Returns the attribute value as a `double?`, or `null` if the element is `null`,
  /// the attribute does not exist, or its value is not a valid double.
  static double? getAttributeValueAsDouble(
      XmlElement? element, String attributeName) {
    if (element == null) return null;
    final attributeValue = element.getAttribute(attributeName);
    if (attributeValue == null) return null;
    return double.tryParse(attributeValue);
  }

  /// Gets the value of an attribute by [attributeName] from an [element] and parses it as an integer.
  ///
  /// Returns the attribute value as an `int?`, or `null` if the element is `null`,
  /// the attribute does not exist, or its value is not a valid integer.
  static int? getAttributeValueAsInt(
      XmlElement? element, String attributeName) {
    if (element == null) return null;
    final attributeValue = element.getAttribute(attributeName);
    if (attributeValue == null) return null;
    return int.tryParse(attributeValue);
  }
}

File: MusicXMLParser/Utils/XmlHelper.cs
using System.Xml.Linq;
using System.Linq;
using MusicXMLParser.Exceptions; // Assuming MusicXmlStructureException is in this namespace
using System.Text.RegularExpressions; // For regex in FindOptionalTextElement
using System.Xml; // For IXmlLineInfo

namespace MusicXMLParser.Utils
{
    public static class XmlHelper
    {
        public static int GetLineNumber(XElement? element)
        {
            if (element == null) return -1;

            // Standard way to get line info in System.Xml.Linq
            IXmlLineInfo? lineInfo = element as IXmlLineInfo; // Use 'as' for safer casting
            if (lineInfo != null && lineInfo.HasLineInfo())
            {
                return lineInfo.LineNumber;
            }

            // Fallback for non-standard 'line' attribute (as in Dart version)
            var lineAttribute = element.Attribute("line")?.Value;
            if (lineAttribute != null && int.TryParse(lineAttribute, out int lineNum))
            {
                return lineNum;
            }
            return -1;
        }

        public static string? FindOptionalTextElement(XElement? element, string path)
        {
            if (element == null || string.IsNullOrEmpty(path)) return null;

            XElement? currentContextNode = element; // currentContextNode can become null
            var pathSegments = path.Split('/');

            for (int i = 0; i < pathSegments.Length; i++)
            {
                if (currentContextNode == null) return null;
                string segment = pathSegments[i];

                if (segment.StartsWith("@"))
                {
                    if (i == pathSegments.Length - 1)
                    {
                        var attributeName = segment.Substring(1);
                        return currentContextNode.Attribute(attributeName)?.Value;
                    }
                    else
                    {
                        return null; // Attribute selection in middle of path not supported
                    }
                }
                else
                {
                    currentContextNode = FindElementFromSegment(currentContextNode, segment);
                }
            }
            return currentContextNode?.Value.Trim();
        }

        private static XElement? FindElementFromSegment(XElement? parent, string segment)
        {
            if (parent == null) return null; // Added null check

            var predicateMatch = Regex.Match(segment, @"(.+?)\[(.+?)\]");

            if (predicateMatch.Success)
            {
                string elementName = predicateMatch.Groups[1].Value;
                string predicate = predicateMatch.Groups[2].Value;

                var attributePredicateMatch = Regex.Match(predicate, @"@(.+?)=""(.+?)""");
                if (attributePredicateMatch.Success)
                {
                    string attributeName = attributePredicateMatch.Groups[1].Value;
                    string attributeValue = attributePredicateMatch.Groups[2].Value;
                    return parent.Elements(elementName)
                                 .FirstOrDefault(el => el.Attribute(attributeName)?.Value == attributeValue);
                }
                // Fallback for unhandled or malformed predicate
                return parent.Elements(elementName).FirstOrDefault();
            }
            else
            {
                return parent.Elements(segment).FirstOrDefault();
            }
        }

        public static XElement GetRequiredElement(XElement parent, string name, string? requiredElement = null)
        {
            // parent is assumed to be non-null by contract of this method.
            // If parent could be null, an ArgumentNullException or different handling is needed.
            var elements = parent.Elements(name);
            if (!elements.Any())
            {
                throw new MusicXmlStructureException(
                    $"Required element <{requiredElement ?? name}> not found as a child of <{parent.Name.LocalName}>.",
                    requiredElement ?? name,
                    parent.Name.LocalName,
                    GetLineNumber(parent),
                    null // context can be null here
                );
            }
            return elements.First();
        }

        public static XElement? FindOptionalElement(XElement? parent, string name)
        {
            return parent?.Elements(name).FirstOrDefault(); // Use null-conditional access
        }

        public static int? GetElementTextAsInt(XElement? element)
        {
            if (element == null) return null;
            var text = element.Value.Trim();
            return int.TryParse(text, out int result) ? result : (int?)null;
        }

        public static double? GetElementTextAsDouble(XElement? element)
        {
            if (element == null) return null;
            var text = element.Value.Trim();
            // Use System.Globalization.CultureInfo.InvariantCulture for consistent parsing
            return double.TryParse(text, System.Globalization.NumberStyles.Any, System.Globalization.CultureInfo.InvariantCulture, out double result) ? result : (double?)null;
        }

        public static bool GetElementTextAsBool(XElement? element, bool defaultValue = false)
        {
            if (element == null) return defaultValue;
            var text = element.Value.Trim().ToLowerInvariant();
            if (text == "yes") return true;
            if (text == "no") return false;
            return defaultValue;
        }

        public static string? GetAttributeValue(XElement? element, string attributeName)
        {
            return element?.Attribute(attributeName)?.Value;
        }

        public static double? GetAttributeValueAsDouble(XElement? element, string attributeName)
        {
            var attributeValue = element?.Attribute(attributeName)?.Value;
            if (attributeValue == null) return null;
            // Use System.Globalization.CultureInfo.InvariantCulture for consistent parsing
            return double.TryParse(attributeValue, System.Globalization.NumberStyles.Any, System.Globalization.CultureInfo.InvariantCulture, out double result) ? result : (double?)null;
        }

        public static int? GetAttributeValueAsInt(XElement? element, string attributeName)
        {
            var attributeValue = element?.Attribute(attributeName)?.Value;
            if (attributeValue == null) return null;
            return int.TryParse(attributeValue, out int result) ? result : (int?)null;
        }

        public static bool? GetAttributeValueAsBool(XElement? element, string attributeName)
        {
            var attributeValue = element?.Attribute(attributeName)?.Value;
            if (attributeValue == null) return null;
            return attributeValue.ToLowerInvariant() switch
            {
                "yes" => true,
                "no" => false,
                _ => null
            };
        }

        public static string? GetElementText(XElement? parent, string elementName)
        {
            return parent?.Element(elementName)?.Value?.Trim();
        }

        public static int? GetElementTextAsInt(XElement? parent, string elementName)
        {
            var element = parent?.Element(elementName);
            return GetElementTextAsInt(element);
        }

        public static double? GetElementTextAsDouble(XElement? parent, string elementName)
        {
            var element = parent?.Element(elementName);
            return GetElementTextAsDouble(element);
        }

        public static bool GetElementTextAsBool(XElement? parent, string elementName, bool defaultValue = false)
        {
            var element = parent?.Element(elementName);
            return GetElementTextAsBool(element, defaultValue);
        }

        public static bool HasElement(XElement? parent, string elementName)
        {
            return parent?.Element(elementName) != null;
        }

        public static int GetElementCount(XElement? parent, string elementName)
        {
            return parent?.Elements(elementName).Count() ?? 0;
        }

        public static string? FindOptionalTextElementOptimized(XElement? element, string path)
        {
            if (element == null || string.IsNullOrEmpty(path)) return null;

            if (!path.Contains('/') && !path.Contains('['))
            {
                return element.Element(path)?.Value?.Trim();
            }

            return FindOptionalTextElement(element, path);
        }
    }
}
